x/authz/keeper/keeper.go

```diff
@@ -113,6 +113,13 @@ func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []
 		if err != nil {
 			return nil, sdkerrors.Wrapf(err, "failed to execute message; message %v", msg)
 		}
+		// emit the events from the dispatched actions
+		events := msgResult.Events
+		sdkEvents := make([]sdk.Event, 0, len(events))
+		for i := 0; i < len(events); i++ {
+			sdkEvents = append(sdkEvents, sdk.Event(events[i]))
+		}
+		ctx.EventManager().EmitEvents(sdkEvents)
```

Comment:
PR looks good to me, but I actually don't understand why the underlying Msg's events are not emitted in the first place?

L112 we call `msgResult, err = handler(ctx, msg)`, that that should call `ctx.EventManager.EmitEvents()` already.

Reply:
I think it's because MsgExec runs its tx messages directly through tendermint/ABCI rather than the SDK handlers? Thats the gist i got from digging into it

Reply:
Let's make sure we dig through the logic in the router and BaseApp and understand what's happening well

Reply:
Checked some more and looks like this [code](https://github.com/cosmos/cosmos-sdk/blob/master/baseapp/msg_service_router.go#L110) overwrites the current event manager in context, which i assume could be a culperit here. I added a nil check before allowing that code to run (it was never nil in my tests). 

Curiously though, it duplicated the attributes in the events. Haven't had much luck figuring out why that happens though

Reply:
If we removed that line in `msg_service_router.go`, and reverted all changes in this PR, does that still emit duplicates?

It seems to me that would be a better fix.

Reply:
yeah when i ran my test i had commented out the event emission i added in `DispatchEvents`. For some messages though, not everything was simply a duplicate, for instance MsgExec's event only had duplicated the `sender` attribute, but nothing else.

also looking at this again, im noticing not every event has an exact duplicate - for example the transfer event has 2 different recipient events, however sender and amount are both duplicate copies.